### PR TITLE
Upgrade pytorch to 2.6, python to 3.12

### PR DESCRIPTION
**Pull request by ebr**
## Summary

- **upgrade pytorch to 2.6.0 and unpin some of the strict dependency pins to facilitate upgrading co-dependencies.** Some depenencies are now unpinned to help continuously stay up to date; uv.lock will be used for reproducibility.
- **allow python 3.12 as we're no longer constrained by the torch version**
- **use uv.lock to pin dependencies** - this will help with reproducible builds
- **refactor Dockerfile; get rid of multi-stage build; upgrade to python 3.12**. This should make the images smaller, the builds faster, and could potentially allow us to utilize docker layer cache in GHA. Notably, changing any application code will only update only the very last layer, and will not result in rebuilding of the entire virtualenv, which means with proper caching changes between images will be tiny and builds will be super fast
- **unpin timm and controlnet-aux** - see also https://github.com/invoke-ai/InvokeAI/pull/7846 - may need to rebase this PR if that one is merged first
- **add python 3.12 to the test matrix, update docs to recommend python 3.12 and use `uv` to manage it**
- **remove some obsoleted depenencies that were used by the CLI**
- **reintroduce GPU_DRIVER build arg in CI container build, as it has apparently been removed**, meaning our "rocm" and "cpu" docker images were just cuda images for some time 

## QA Instructions

- Usual manual testing - these changes have a large surface area, so we'll rely on the community to help test all edge cases.
- special attention should be given to testing the Docker container.

## Checklist

- [x] _The PR has a short but descriptive title, suitable for a changelog_
- [x] _Tests added / updated (if applicable)_
- [x] _Documentation added / updated (if applicable)_
- [ ] _Updated `What's New` copy (if doing a release after this PR)_

### CODE REVIEW

**Review by hipsterusername — APPROVED** [review]

**Review by jazzhaiku — APPROVED** [review]
LGTM 👍

File: pyproject.toml [hunk 1]
```diff
@@ -33,39 +33,39 @@ classifiers = [
...
-  "huggingface-hub==0.26.1",
-  "pydantic-settings==2.2.1",
-  "pydantic==2.7.2",
-  "python-socketio==5.11.1",
-  "uvicorn[standard]==0.28.0",
+  "fastapi-events",
+  "fastapi",
+  "huggingface-hub",
+  "pydantic-settings",
+  "pydantic<2.9",
```
Comment (psychedelicious) [inline pyproject.toml:66]: Pydantic is currently at v2.10.6 stable, any reason in particular to pin it below that?

**Comment by psychedelicious** [comment]
How do we regenerate `uv.lock` and how does it interact with pins in `pyproject.toml`?

**Comment by psychedelicious** [comment]
We need to be careful about this change and its interaction w/ launcher.

Though the launcher supports "pinning" versions of invoke to different versions of python, that information is hardcoded and we'll need to do a launcher release.

I suspect attempting to install the version of Invoke that enforces python 3.12 without updating the launcher beforehand will result in errors. 

Maybe we should figure out how to make the launcher query PyPI to get the python version required by a package?

File: pyproject.toml (hunk 1 above)
Comment (ebr) [reply pyproject.toml:66]: yes - the issue we saw with "unbound input" (i can't recall the exact wording) warnings in the console when debugging the LLaVA node were caused by pydantic > 2.9. I did not dig into troubleshooting that, but it would be the next thing to address.

**Comment by ebr** [comment]
> How do we regenerate `uv.lock` and how does it interact with pins in `pyproject.toml`?

`uv lock` re-runs dependency resolution and updates the `uv.lock` file. pins in pyproject.toml still guide the resolution process and can be as restrictive or loose as we want them to be, but the lockfile pins dependencies to the actual hashes in the index. Works the same way as `package-lock.json`.

we do still need to test how `uv.lock` interacts with our use of different indexes, such as for `rocm` and cpu-only pytorch. if there's breakage, there's a "conflicting extras" syntax we might be able to use to work around it:  https://docs.astral.sh/uv/reference/settings/#conflicts

**Comment by ebr** [comment]
> We need to be careful about this change and its interaction w/ launcher.
100%

> Maybe we should figure out how to make the launcher query PyPI to get the python version required by a package?

depends on whether we want to continue supporting <=3.11 ... but yes, the launcher could easily query pypi for this info: 

```
$ curl -s https://pypi.org/pypi/invokeai/json |jq '.info.requires_python'
"<3.12,>=3.10"
```

we can also query for any particular release `| jq '.releases` where `releases` is a mapping of ` vX.Y.Z : [ { <pkg_info> }, {...}  ]`

File: pyproject.toml (hunk 1 above)
Comment (psychedelicious) [reply pyproject.toml]: I've fixed the issue caused by the pydantic version bump and updated this pr accordingly

**Comment by psychedelicious** [comment]
I believe the remaining failing `python tests` CI failures will be resolved by #7826. Some magic pydantic bs.

**Comment by psychedelicious** [comment]
Rebased on `main` & realized I didn't update the uv lockfile correctly before, which caused the frontend type check CI workflow to fail. Fixed that now

**Comment by psychedelicious** [comment]
I believe this PR is ready. At least, all CI is passing. We need to check on how it might interact with the launcher and existing installs. So I've marked it as draft to prevent a premature merge.

**Comment by psychedelicious** [comment]
Superseded by #7873, which I needed to create as a branch on the origin repo to do testing of CI changes. I did quite a bit of extra work on it.
